PolicyOptimization/DistribPolicyGradients/ClientTrajectoryFinalizer.py

```diff
@@ -0,0 +1,148 @@
+import time
+
+import torch
+from Distrib import RedisClient, RedisKeys
+from Distrib.RedisServer import RedisServer
+from MPFramework import MPFProcess
+import numpy as np
+from Policies import PolicyFactory
+from Utils import CompressionSerialisation as cser
+
+
+class ClientTrajectoryFinalizer(MPFProcess):
+    HEADER_INITIALIZATION="initialization"
+    HEADER_FLUSH = "flush_data"
+    HEADER_TRAJECTORY = "trajectory"
+
+    def __init__(self, loop_wait_period=0.1):
+        super().__init__("ClientTrajectoryFinalizer", loop_wait_period=0.1, process_all_updates=True)
+        self.handlers = {
+            ClientTrajectoryFinalizer.HEADER_INITIALIZATION: self._init,
+            ClientTrajectoryFinalizer.HEADER_FLUSH: self._flush,
+            ClientTrajectoryFinalizer.HEADER_TRAJECTORY: self._trajectory
+        }
+    
+    def init(self):
+        self.client = None
+        self.cfg = None
+        self.gamma = None
+        self.lmbda = None
+        self.value_estimator = None
+        self.reward_stats = None
+
+        self.trajectories_to_send = []
+        self.total_timesteps = 0
+        self.t0 = None
+        
+    def update(self, header, data):
+        handler = self.handlers.get(header, None)
+        if not handler:
+            print(f"WARNING: ClientTrajectoryFinalizer received unknown message header '{header}'")
+            return
+
+        handler(data)
+
+    def _init(self, cfg):
+        self.cfg = cfg
+        self.gamma = self.cfg["policy_optimizer"]["gamma"]
+        self.lmbda = self.cfg["policy_optimizer"]["gae_lambda"]
+
+        pf_cfg = {
+            "device": cfg.get("device", "cpu"),
+            "value_estimator": cfg["value_estimator"]
+        }
+
+        models = PolicyFactory.get_from_cfg(pf_cfg, env_space_shapes=cfg["env_space_shapes"])
+        self.value_estimator = models["value_estimator"]
+
+        if self.client is not None:
+            try:
+                self.client.disconnect()
+            except:
+                pass
+
+        self.client = RedisClient()
+        self.client.connect()
+
+        value_params = None
+        while value_params is None:
+            value_params = self.client.get_latest_value_params()
+
+        print(f"updated value estimator to version {self.client.current_value_epoch}")
+
+        self.value_estimator.set_trainable_flat(value_params)
+        self.reward_stats = self.client.get_reward_stats()
+
+        self.trajectories_to_send = []
+        self.total_timesteps = 0
+    
+    def _flush(self, rewards):

```

Comment:
Probably should call out the fact that this sends rewards, if available. Ideally this would happen in a separate message first, then I'd just `flush` w/o a `data` arg.

Reply:
I left this as-is for now, as it's more efficient that way.